Implement UserAnalytics

brecriffs opened the pull request:
#### Summary
This pull request contains changes made to the server to accommodate requests from the front-end for user interactions data.

**api4/user.go**
- Added api get route ‘users/{userId}/teams/{teamId}/useranalytics‘
    - Requires 1 parameter 
       - name: a string indicating the data set that you want to fetch. The names of data sets are defined in mattermost-redux/src/constants/user_analytics.js (Datasets)

- Added function getUserAnalytics() to handle route
    - Calls app/user.go and eventually returns data to mattermost-redux

**app/user.go**
- Added GetUserAnalytics() function, which calls to the service layer and runs the sql queries, appends the data to a UserAnalytics struct, and finally returns all data back to api4/user.go
    - Currently the only data set that can be explicitly requested is 'UserInteractions'
    - 'UserLearningGroups' data is always returned in the request
    - 'LearningGroupPrefixes' data is always returned in the request
- Added GetLearningGroupPrefixes() function, which given a teamId:
    1. gets data for the current team (most importantly, the slug name)
    2. finds the LMS in the config that has a team defined with that slug name
    3. returns the learning group prefixes for that LMS (using the GetLearningGroupPrefixes utility function in model/edx.go)

**model/edx.go**
- Added GetTeams() function, which given a learning management service object (LMS), returns the teams and their slug names from the config file
- Added GetLearningGroupPrefixes() function, which given a learning management service object (LMS), returns the keys to the ChannelList object in that LMS
- Added CheckIfActiveLMS() function, which given a learning management service object (LMS) and team slug, determines if that team slug exists in the LMS, which indicates that the LMS is the active LMS 

**store/sqlstore/user_store.go**
- Added GetUserInteractions() function, which runs sql query and returns data back to app/user.go
    - User interaction data is returned as an array of pointers to UserInteraction structs, where each struct represents a single user interaction. The struct is defined in (/model/user.go), with the following properties:
        - _Interaction_type_
            - ‘Reaction’
            - ‘Mention’
            - ‘Reply’
            - ‘DirectMessage'
        - _Username_
            - Username of the other user in a row’s interaction
            - Could be the recipient or the creator of a row’s interaction
        - _Is_Recipient_	
            - Boolean denoting whether the current user is the recipient of a row’s interaction
        - _Context_
            - ‘Course’
        - _Channel.Type_ 
            - Useful for debugging and implementation on front-end
            -Will be a char, denoting the type of channel that the interaction took place in, one of:
                - P -private channel
                - D -direct message
                - G -group message
                - O -public channel
        - _Channel.Name_ 
            - Useful for debugging and implementation on front-end
            - The name of the channel that the interaction took place in (ex. town-square)
- Added GetUserLearningGroups() function, which runs sql query and returns data back to app/user.go
    - User learning group data is returned as an array of pointers to LearningGroup structs, with a struct for each learning group that the user is a member of (ex. PLG or Capstone). The struct is defined in (/model/user.go), with the following columns:
        - _Channel_Id_
            - id of the channel for this team
        - _Channel_Name_
            - name of the channel for this team (ex. plg-30)
        - _Members_
            - a comma delimited list of the usernames of the users that are in this team's channel
        - _Team_Name_
            - the Channel_Name without the initial team type (ex. plg-30 => '30')
        - _Team_Type_
            - the type of team (ex. plg-30 => 'plg')
       - _Has_Left_Group_
           - bool denoting whether or not the user has left his learning group

**store/storetest/mocks/UserStore.go**
_Need to look further into how/why these are used...the console had an error if this wasn’t present. They map to functions in 'store/sqlstore/user_store.go'_
- Added GetUserInteractions() function, which is a mock function for mattermost tests
- Also added GetUserLearningGroups() function, for same reason

#### Ticket Link
https://app.zenhub.com/workspaces/riff-projects-5cb6029b4103d93ceadf001b/issues/rifflearning/zenhub/101

brecriffs commented:
This is ready for review. Upon further QA of the contact counts, the SQL query may have to be adjusted slightly, but in general, I am happy with the structure of the code.

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
 	ReturnStatusOK(w)
 }
+
+func getUserAnalytics(c *Context, w http.ResponseWriter, r *http.Request) {
```
I'd prefer to have a documenting comment attached to this function. I don't care too much about the format. Ideally, it would include the following (I've listed them in order of importance):

- A single sentence statement of the function's purpose.
- A single line description of each parameter.
- A statement of the preconditions and postconditions associated with calling the functions. What invariant is expected to be true before the function executes? What has changed about the world after the execution is complete?  

Also, the variable names here (c, w, and r) are maybe not descriptive enough? Their types are pretty descriptive, but maybe using self-documenting names would be an alternative to adding a documenting comment.

brecriffs replied:
Added some documentation for this function. I think we should keep the variable names (c, w, r), to keep this function consistent with other functions. 

Also added more error handling for parameters that are passed. Now handles null, undefined, false, "", or invalid teamdId's and userId's gracefully, with an informative error printed to the server console. (Should have been done in the first place)

mlippert replied:
Consistency w/ the existing codebase is very important. (Although there is very little in the way of documentation in the current codebase and _that_ I think we can and should do better!)

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
+	// Data sets are mirrored in the struct, UserAnalytics, in model/user.go
+	types := r.URL.Query().Get("types")
+
+	// The id of the current mattermost team
```
A lot of these single-line comments seem like they could be removed and replaced by names that are more self-documenting. In general, I favor self-documenting names over a separate comment like this because someone who changes the function of the code is more likely to notice that the variable is no longer being used for its stated purpose if the purpose _is_ the name of the variable.

Here, consider using the name currentMattermostTeamId perhaps?

brecriffs replied:
I think once you get to the mattermost-server repo, teamId is ubiquitous and seems obvious that it is the id of the current mattermost team. I think the variable name is fine, if anything the comment is redundant. I just figured I would follow convention, but also comment everything for clarity, even if the naming convention or code style is used all over the repo.

jaedoucette replied:
If the name is used throughout the code base, it makes sense to keep the convention.

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
+	// The id of the current mattermost team
+	teamId := r.URL.Query().Get("teamId")
+
+	// The user id of the logged in user
```
currentUserID? or loggedInUserID?

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
+	// The user id of the logged in user
+	uid := r.URL.Query().Get("uid")
+
+	// Call to service layer (/app/user.go)
```
This comment seems more helpful to me than the other ones.

brecriffs replied:
This prompted me to look more closely at how other user related routes handle api requests. Also, after peeking at some comments by @mlippert on the mattermost-redux PR, I made some changes to how the route is handled for the useranalytics request, in anticipation of and to accommodate changes to the mattermost-redux repo.

I no longer use the variable uid. I now use userId, which seems to be the naming convention (besides a few cases) in this repo. Also, the userId is no longer passed as a parameter, and is received here using `c.Params.UserId`, which also seems to be a convention.

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
+	}
+
+	// Return user analytics data to mattermost-redux as json
+	w.Write([]byte(rows.ToJson()))
```
General question: I would have guessed that this http ResponseWriter would have some exceptional failure modes, but it doesn't appear to in the GoLang documentation. Is that really so? Kinda nice.

brecriffs replied:
What do you mean by 'exceptional failure modes'? You mean like, there aren't any concerns about errors on the line w.Write([]byte(rows.ToJson()))?

jaedoucette replied:
I mean literally that byte writing would usually be something that could throw an exception, perhaps because some of the elements of `rows` can't be properly serialized to JSON, say. The documentation for `ResponseWriter.Write(...)` doesn't mention any such possibility, so probably it's fine, but I was surprised not to see logic for handling possible exceptions.

brecriffs replied:
I think I did such a thorough job of type checking throughout, and then defined a function explicitly for converting that type into JSON, but I agree, surprised other functions don't implement an error catch on that statement.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
 	return nil
 }
+
+func (a *App) GetUserAnalytics(uid string, teamId string, types string) (*model.UserAnalytics, *model.AppError) {
```
Per comment in user.go, would be nice to have a documenting comment at the start of this function. Ideally, maybe think about using a commenting style that is compatible with some kind of auto-doc. It looks like Go has a standardized approach here: https://blog.golang.org/godoc-documenting-go-code

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
 }
+
+func (a *App) GetUserAnalytics(uid string, teamId string, types string) (*model.UserAnalytics, *model.AppError) {
+	// Initialize returnData struct using UserAnalytics model
```
What is returnData exactly? I get that this is documented elsewhere (probably in the UserAnalytics model's declaration), but I think either drop the comment, or provide a clearer explanation of what the UserAnalytics structure will represent here.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	// Initialize returnData struct using UserAnalytics model
+	returnData := new(model.UserAnalytics)
+
+	// Instantiate learningGroupPrefixes and error structs
```
Ditto here. This comment doesn't tell me a whole lot more than what reading the code would tell me. I like the idea of commenting declarations like this, but I think having either more self-documentation in the names, or a comment that tells me more about what these variables are for, would be clearer.

brecriffs replied:
Added a more informative comment. I think returnData is an informative enough variable name, especially when setting the type UserAnalytics. It makes me think "this is the return data for user analytics". What do you think?

jaedoucette replied:
This comment was meant to be about the next couple of declarations `learningGroupPrefixes` and `err`. I have another one above about the `returnData` comment.

At least for me, the type `UserAnalytics` doesn't tell me much. A comment reminding me what that type is supposed to represent seems like it could be helpful here.

brecriffs replied:
Added some more comments in this area of code

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	var learningGroupPrefixes []*model.LearningGroupPrefix
+	var err *model.AppError
+
+	// Get learning group prefixes (ex. 'plg' or 'capstone')
```
This makes things quite a bit clearer to me. If you told me at the declaration of learningGroupPrefixes that it was to store the prefix portion of channel names, which by convention in our courses, is used to indicate whether groups were created at the start of the course, or later in the course, then I think this would be super clear for a future reader.

brecriffs replied:
Added a more informative comment. How I am using the prefixes is primarily to reasonably assert that a private channel was automatically generated for a learning group, and I've documented that more clearly.

jaedoucette replied:
Much clearer!

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	var err *model.AppError
+
+	// Get learning group prefixes (ex. 'plg' or 'capstone')
+	if learningGroupPrefixes, err = a.GetLearningGroupPrefixes(teamId); err != nil {
```
This is a neat idiom. Since `err` is never used outside the if though, I wonder if it would be more idiomatic to declare it as a local variable in the initialization?

brecriffs replied:
Yeah this syntax definitely looked weird to me at first, but it seems to be used throughout the repo as a quick kind of shortcut.

brecriffs replied:
Oh I think I originally misunderstood what you meant. What do you mean by 'declare it as a local variable in the initialization'? I could have done `if learningGroupPrefixes, err :=` , with the shorthand variable declaration (:=), and not initialise learningGroupPrefixes or err, but I wanted to ensure type by declaring learningGroupPrefixes as type '[]*model.LearningGroupPrefix'.

jaedoucette replied:
Yeah, that's what I meant, but I wasn't sure if there was a clean way to declare + initialize `err` locally and simultaneously initialize the `learningGroupPrefixes` variable that was declared in the outer scope. If there isn't, then the current code is fine, but if there is, it seems like that might be nicer.

brecriffs replied:
You can do that, but only by using the shorthand method described above, and the learningGroupPrefixes variable wouldn't get initialised as type `[]*model.LearningGroupPrefix`.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+
+	// Add learningGroupPrefixes to return data
+	// Assert as type 'array of pointers to LearningGroupPrefix structs'
+	returnData.LearningGroupPrefixes = interface{}(learningGroupPrefixes).([]*model.LearningGroupPrefix)
```
It _might_ be preferable here to perform a test rather than a cast? From my reading, golang's type assertion will trigger an unhanded exception runtime exception here if the types don't match. I think it would be better practice to use the 2-return value form of this (i.e. test, don't cast), which gives you a return value to check whether the cast succeeded, and then perform a controlled exit or display an informative logging message before panic?

jaedoucette commented on api4/user.go:
```diff
@@ -1568,3 +1570,29 @@ func registerTermsOfServiceAction(c *Context, w http.ResponseWriter, r *http.Req
+	// Types of data sets to return
+	// All data sets are defined as Datasets in mattermost-redux/src/constants/user_analytics
+	// Data sets are mirrored in the struct, UserAnalytics, in model/user.go
+	types := r.URL.Query().Get("types")
```
Type of `types` is unclear here. Later, types is primarily used to perform `contains` operations. I think it is a list of strings based on my reading of the docs.  Depending on whether we want to make sure this scales, it might be preferable to construct a set out of this list instead?

brecriffs replied:
I originally pass it as an array, but it gets converted into a comma delimited list of strings during the query string building. I agree using .contains on a string kind of sucks. I now convert it back to an array of strings, and use the `utils.StringInSlice` function to check if the array contains a certain data set. I think that is better. What do you think?

jaedoucette replied:
Since StringInSlice is linear time, I think the scaling advantage over contains is probably small. Better would be to build a set out of the list, which will have constant- or log-time contains. This is only true if the list will be "large" however, either now or in future. It's not clear to me what types actually is, so I'm not sure it will be large enough to merit this optimization.

brecriffs replied:
Ah makes sense. Is a set like an object? We are now just passing a 'name' variable for identifying one data set to return, so this has become redundant.

jaedoucette replied:
A set is a data structure. It's a bit like a list, except that:

- Items must be unique (usually). If you try to put a duplicate in, it usually isn't added.
- To determine whether something is in a set is (usually) O(1) or O(log n) time (where n is the number of elements), whereas for a list it is O(n) time.

In this application, where you do O(n) contains queries, you get O(n) total runtime with a set, but O(n^2) with a list. For a list of a 1000, that would be a million extra ops, but for a list of 100, it'd be only 9000 extra ops. That's why I wasn't sure whether this optimization was worth thinking about.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	returnData.LearningGroupPrefixes = interface{}(learningGroupPrefixes).([]*model.LearningGroupPrefix)
+
+	// Manually check if types contains each graph type
+	// types is a comma delimited string converted from an array during the api request
```
So here it looks like types is a string, not a list of strings. Probably this is fine for now, but it would be more salable for sure to do a split and then build a set from the resulting list.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+		if userInteractions := <-a.Srv.Store.User().GetUserInteractions(uid, teamId, learningGroupPrefixes); userInteractions.Err != nil {
+			// Return and exit from GetUserAnalytics with error
+			return nil, userInteractions.Err
+		} else {
```
idiomatic golang appears to prefer avoiding else after if + return.

brecriffs replied:
This syntax seems to be used throughout the repo, especially for `a.Srv.Store.User().` functions. I think this is fine.

brecriffs replied:
As discussed I think there are scoping issues when not using the else AND using the shorthand variable declaration.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	// Manually check if types contains each graph type
+	// types is a comma delimited string converted from an array during the api request
+	if strings.Contains(types, "UserInteractions") {
+		if userInteractions := <-a.Srv.Store.User().GetUserInteractions(uid, teamId, learningGroupPrefixes); userInteractions.Err != nil {
```
Line too long. Maybe try https://godoc.org/golang.org/x/lint

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	}
+
+	// Always return the user's learning groups
+	if userLearningGroups := <-a.Srv.Store.User().GetUserLearningGroups(uid, teamId, learningGroupPrefixes); userLearningGroups.Err != nil {
```
Line too long?

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	if userLearningGroups := <-a.Srv.Store.User().GetUserLearningGroups(uid, teamId, learningGroupPrefixes); userLearningGroups.Err != nil {
+		// Return and exit from GetUserAnalytics with error
+		return nil, userLearningGroups.Err
+	} else {
```
else after return

brecriffs replied:
This syntax seems to be used throughout the repo, especially for a.Srv.Store.User(). functions. I think this is fine.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+		} else {
+			// Add UserInteractions data to the returnData object
+			// Assert as type 'array of pointers to UserInteraction structs'
+			returnData.UserInteractions = userInteractions.Data.([]*model.UserInteraction)
```
cast without check

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	} else {
+		// Add LearningGroups data to the returnData object
+		// Assert as type 'array of pointers to LearningGroup structs'
+		returnData.UserLearningGroups = userLearningGroups.Data.([]*model.LearningGroup)
```
cast without check.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	return returnData, nil
+}
+
+func (a *App) GetLearningGroupPrefixes(teamId string) ([]*model.LearningGroupPrefix, *model.AppError) {
```
Documenting comment for function declaration

brecriffs replied:
Added documenting comment

jaedoucette replied:
Much clearer!

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	var currentTeam *model.Team
+	if result := <-a.Srv.Store.Team().Get(teamId); result.Err != nil {
+		return nil, result.Err
+	} else {
```
else after return.

brecriffs replied:
This syntax seems to be used throughout the repo, especially for `a.Srv.Store.Team().` functions. I think this is fine.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+		currentTeam = result.Data.(*model.Team)
+	}
+
+	// Instantiate learningGroupPrefixes struct
```
Comment doesn't add meaning beyond existing code.

brecriffs replied:
removed comment

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	// Instantiate learningGroupPrefixes struct
+	var learningGroupPrefixes []*model.LearningGroupPrefix
+
+	// For every object in the LMSs array in the config
```
I think this comment should document the _ and lms? I guess we're iterating over a list of pairs?
I'd rather have a single block comment at the start of this nested control structure telling me the overall purpose, instead of interleaving them.

brecriffs replied:
I like the comments throughout, but I did improve them in the commit 'Improved documentation for looping over configuration objects'.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+
+	// For every object in the LMSs array in the config
+	for _, lms := range a.Config().LTISettings.GetKnownLMSs() {
+			// For every property of Teams in this learning management service (LMS)
```
It would be more helpful here to have an explanation of what a teamSlug is.

The comment and code together are actually a bit confusing to me. The code looks like we are iterating over teams (perhaps it's pairs of teams and teamSlug properties?), but the comment suggests we are iterating over _properties_ of a `Teams` structure? More detail would definitely help here, or more self-documenting names.

brecriffs replied:
See commit 'Improved documentation for looping over configuration objects'.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+	for _, lms := range a.Config().LTISettings.GetKnownLMSs() {
+			// For every property of Teams in this learning management service (LMS)
+			for _, teamSlug := range lms.GetTeams() {
+				// If this property is equal to the user's currently selected mattermost team slug
```
Here, maybe "the current user's currently selected mattermost team slug"?

brecriffs replied:
See commit 'Improved documentation for looping over configuration objects'.

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,78 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+
+	// Return array of pointers to LearningGroupPrefix structs and nil as error
+	// Assert learningGroupPrefixes as type 'array of pointers to LearningGroupPrefix structs'
+	return interface{}(learningGroupPrefixes).([]*model.LearningGroupPrefix), nil;
```
Cast without test?

brecriffs replied:
Added tests to handle assertion errors.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+	return e.Teams
+}
+
+func (e *EdxLMS) GetLearningGroupPrefixes() []*LearningGroupPrefix {
```
Documentation for function declaration missing

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+}
+
+func (e *EdxLMS) GetLearningGroupPrefixes() []*LearningGroupPrefix {
+	// Instantiate array of pointers to LearningGroupPrefix structs
```
Comment doesn't add information beyond the information in the code.

Also, `size` is never used outside of the call to `make`. It seems like it would be cleaner to pass `len(...)` directly instead of creating the local variable? Having `size` as a local that you then use in a dynamic allocation suggests to me that you are going to try to do some more complicated manual memory allocation later using `size`, but this doesn't happen.

brecriffs replied:
Makes sense, removed redundant comment and variable declaration

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+	var size int = len(e.PersonalChannels.ChannelList)
+	var prefixes []*LearningGroupPrefix = make([]*LearningGroupPrefix, size)
+
+	// For every key in PersonalChannels.ChannelList in the active configuration
```
I think this comment isn't adding anything beyond what's in the code. Either remove or explain the whole behaviour of the loop here?

brecriffs replied:
I'm kind of used to documenting for loops like this. I iteratively comment each part of a loop to say what it's doing. I find it easier to see what's happening than having a large comment at the beginning, then trying to match parts of the comment to parts of code. 

ex. 
```
var fruits = ['banana', 'apple', 'orange']
var bowl = [];

// For every fruit
for (let i = 0; i < fruits.length; i++) {
  // Add the fruit to the bowl
  bowl.push(fruits[i]);
}
```



What do you think @mlippert?

jaedoucette replied:
I think in general I'd be okay with that, and in the example you show, that style makes sense.

In the code though, you're using a for each loop, not a for loop, so it reads more like:

`var fruits = ['banana', 'apple', 'orange']
var bowl = [];

// For every fruit
for  fruit in fruits {
  // Add the fruit to the bowl
  bowl.push(fruits[i]);
}`

The comment at the top is thus not very helpful, since it reads almost identically to the code.

For me at least, the syntax of the language already tells me _that_ you are putting the fruits in the bowl.  It would be more helpful to have something like this, which tells me _why_ you wanted to put the fruits in the bowl. That's more useful for someone coming after, because maybe GoLang 8 released a new SuperBowl structure that does this better, or maybe there's a refactor to remove the function call this bowl was constructed for.

var fruits = ['banana', 'apple', 'orange']
var bowl = [];

// First, put all the fruits into a bowl, so that we can take advantage of the bowl's 
// efficient internal ordering and nice API
// or, so that we can call a function that expects a bowl
for (let i = 0; i < fruits.length; i++) {
  // Add the fruit to the bowl
  bowl.push(fruits[i]);
}

brecriffs replied:
Yeah I see your point about the wording, it does imply a 'for each'. Revised the wording, and pulled comments to the beginning of the loop.

mlippert replied:
@brecriffs So I agree w/ @jaedoucette that comments that say exactly what the following line of code do are not helpful.

If the loop is simple then a comment like "Add every type of fruit to the bowl" or "Add one of each fruit to the bowl" or something that describes what you'll have after the loop seems best. If the actions in the loop are more complicated, then a comment about what the loop iterates over "Examine each fruit for bugs and only add fruit that can be kept cold to the bowl" as the loop comment and inside the loop "Get the info on this fruit, and check all the storage conditions for cold" and then "Add cold resistant fruit to bowl" on some code in the loop or perhaps "Skip non cold resistant fruit" depending on the implementation in the loop.

I wouldn't usually want @jaedoucette 's comment about internal ordering and API either unless there's something really special about them. In other words if I were to be putting a bunch of things in a Set object, I would not comment that I'm using a Set in order to reduce the things to a unique list. I'd expect the developer to already understand that as a property of Sets. I might comment "Get all the types of fruit in the bowl" or some such.

jaedoucette replied:
@mlippert Your suggestions on this example make sense to me. I also wouldn't want comments about the API or internal ordering unless the structures were something unusual.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+	var prefixes []*LearningGroupPrefix = make([]*LearningGroupPrefix, size)
+
+	// For every key in PersonalChannels.ChannelList in the active configuration
+	counter := 0
```
I think `counter` could have a more self-documenting name. How about `numberPrefixesStored`?

brecriffs replied:
Yeah, good call. Made that change.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+	counter := 0
+	for personalChannelName, _ := range e.PersonalChannels.ChannelList {
+		// Add the pointer to this LearningGroupPrefix struct to the array of pointers
+		prefixes[counter] = &LearningGroupPrefix{Prefix:personalChannelName}
```
why not `prefixes[counter++]` here? I think either make use of the inline increment semantics, or if we're going to have a separate statement for incrementing, try `counter += 1`, which has more explicit semantics.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+		// Add the pointer to this LearningGroupPrefix struct to the array of pointers
+		prefixes[counter] = &LearningGroupPrefix{Prefix:personalChannelName}
+
+		// Increment counter in loop (using range over key, value types doesn't produce a counter)
```
I'd drop this comment.

brecriffs replied:
Also good call. I think I added the comment because my initial thought looking at this would be, 'doesn't a for loop come with a counter usually?'. But with a more descriptive variable name for the counter, I think this comment is safe to remove.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+	}
+
+	// Return array of pointers to LearningGroupPrefix structs
+	return interface{}(prefixes).([]*LearningGroupPrefix)
```
Cast without test?

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,29 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+		counter++;
+	}
+
+	// Return array of pointers to LearningGroupPrefix structs
```
Comment doesn't add clarity.

brecriffs replied:
Fair enough, removed.

jaedoucette commented on model/user.go:
```diff
@@ -47,6 +47,38 @@ const (
 	USER_PASSWORD_MAX_LENGTH  = 72
 )
 
+// This is a struct for a user interaction
```
Either more documentation about what this is supposed to represent (what _is_ a user interaction?), or omit. 

Maybe consider documenting each field of the struct in a comment here.

jaedoucette replied:
Also, I'm guessing the fields are all pointer types to handle the case that the JSON object they represent had a `null` field? This should be documented, because it's not obvious why you'd want, say, a `*bool` here.

brecriffs replied:
Documented all the structs that I added, and their fields, including a note about why I used pointers.

jaedoucette replied:
This is loads clearer.

jaedoucette commented on model/user.go:
```diff
@@ -47,6 +47,38 @@ const (
+	Channel_Name     *string `json:"Channel_Name"`
+}
+
+// This is a struct for a learning group prefix found in the config
```
as above, although this comment is a little clearer.

jaedoucette commented on model/user.go:
```diff
@@ -47,6 +47,38 @@ const (
+	Prefix string
+}
+
+// This is a struct for all user analytics data sets
```
As above.

brecriffs replied:
Improved this.

jaedoucette commented on model/user.go:
```diff
@@ -47,6 +47,38 @@ const (
+	LearningGroupPrefixes []*LearningGroupPrefix
+}
+
+// This is a struct for a learning group (ex. capstone)
```
Ditto.

brecriffs replied:
Improved this.

jaedoucette commented on model/user.go:
```diff
@@ -347,6 +379,13 @@ func (u *UserAuth) ToJson() string {
 }
 
+// This is used in api4/user.go (GetUserAnalytics())
+// ToJson() is called on a UserAnalytics struct pointer before returning data
```
I like this documentation. More like this!

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
 	})
 }
+
+func (us *SqlUserStore) GetUserInteractions(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
```
document function declaration. Also, this line is too long.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+func (us *SqlUserStore) GetUserInteractions(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
+
+	// For every learning group prefix for this team, use this case statement
+	// to record the interaction in this prefix's context (ex. plg)
```
I don't see a case statement here.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+	// to record the interaction in this prefix's context (ex. plg)
+	prefixQuery := ""
+
+	for _, prefix := range learningGroupPrefixes {
```
Consider using a [StringBuilder](https://golang.org/pkg/strings/#Builder) here. This looks like it could get to be a very large string, and its construction is going be quadratic in its length.

jaedoucette replied:
I _think_ this isn't going to be a big deal because there are only 2 prefixes? If I'm wrong and there could be lots, then StringBuilder for sure, to ensure scalability.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+	for _, prefix := range learningGroupPrefixes {
+		prefixQuery += `
+			WHEN Channels.Type = 'P'
+			AND LEFT(Channels.Name,` + strconv.Itoa(len(prefix.Prefix) + 1) + `) = '` + prefix.Prefix + `-'
```
line too long.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+
+	// If there are no learning group prefixes, create negative test case
+	// No prefixes means: for this mattermost team, there are no properties in the
+	//                    ChannelList field in the config for this LMS
```
These 3 comments are pretty helpful.

mlippert replied:
There is no such thing as _this LMS_ at this place in the code. I'd just say it means that "either there is no LMS for this team or that the LMS for this team had no PersonalChannels." Or you could be less pedantic and say "either this team is not associated with a course, or that course had no PersonalChannels." and leave it to the reader to understand that teams are associated with courses via the LTISettings or imply that with a short parenthetical "(see config LTISettings)"

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+			query :=
+			`
+			SELECT * FROM (
```
Consider adding some documentation to this query. Since it's got nested queries, I need to get out paper and pencil to work out what it is doing. If you were able to structure it as a chain of operations (e.g. like in a pipeline or a monad), then I think that would also be very readable without documentation.

jaedoucette replied:
After reading it over, I think this query is maybe too large. It's embedding a lot of business logic directly in the query.

Consider converting to a workflow involving a [query builder](https://github.com/Masterminds/squirrel), or to splitting this into a number of separate smaller queries and doing some of the logical operations in memory instead. I think the query builder approach is most readable in general, but I haven't used a library for it in Go before, so I can't promise that there's a good one. The one I linked looks okay though.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+			var userInteractions []*model.UserInteraction
+
+			if _, err :=  us.GetReplica().Select(&userInteractions, query, map[string]interface{}{"uid": uid, "teamId": teamId}); err != nil {
+				result.Err = model.NewAppError("SqlUserStore.GetUserInteractions", "store.sql_post.ge_user_interactions.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Lines too long here.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+			`
+			var userInteractions []*model.UserInteraction
+
+			if _, err :=  us.GetReplica().Select(&userInteractions, query, map[string]interface{}{"uid": uid, "teamId": teamId}); err != nil {
```
Passing the address of an array here seems like a funny thing to do to me. Presumably this is so that the function can mutate the variable to point at an array that it will allocate? I assume this is fine in Go, but it would be a strange idiom in most C-family languages, so maybe double check.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+	})
+}
+
+func (us *SqlUserStore) GetUserLearningGroups(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
```
document function declaration. Also, line too long.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+func (us *SqlUserStore) GetUserLearningGroups(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
+	// For every learning group prefix for this team, use this case statement
+	// to record a name for this channel(team), minus the prefix (ex. plg-30 => 30)
+	prefixReplaceQuery := ""
```
It looks like there's some boiler-plate here. This makes me even more sure that query builders would be helpful in making this code more readable.

mlippert replied:
I agree that this SQL query is very large and would be better broken into smaller parts.
A query builder might be a good way to go, but I'm loathe to figure out how one correctly adds a package (and figure out the right query builder package to use) in a go project at this time.
What I think would be the compromise is to note this in in a TODO type comment at the top of the query, or in the function definition as something that should be addressed either explicitly or the next time a story requires a change to this function.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+		`
+	}
+
+	// If there are no learning group prefixes, then '' as team name,
```
More boilerplate.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+		prefixReplaceQuery = " WHEN 1 = 1 THEN '' "
+	}
+
+	// For every learning group prefix for this team, use this case statement
```
This is _very_ similar to the loop above. The comment should document what is different. It looks to me like it's just the final line. Again, the boilerplate here suggests you could abstract this loop out as a function and just call it in the 3 places it's being used, along with the associated if.

Also, StringBuilder is looking increasingly like a good idea, since we're doing this quadratic-time string construction in so many places.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+
+	// For every learning group prefix for this team, use this case statement
+	// to find all channels with a name that starts with the prefix
+	prefixFilterQuery := ""
```
As above.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+		prefixFilterQuery = " 1 != 1 "
+	}
+
+	return store.Do(func(result *store.StoreResult) {
```
And here, I'm again convinced that we want a Query builder, or at least a lot of documentation.

jaedoucette commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,466 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+			`
+			var userTeams []*model.LearningGroup
+
+			if _, err :=  us.GetReplica().Select(&userTeams, query, map[string]interface{}{"uid": uid, "teamId": teamId}); err != nil {
```
Lines too long. Same question as above about passing address of array.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
 }
+
+// GetUserInteractions provides a mock function with given fields: uid, teamId, learningGroupPrefixes
+func (_m *UserStore) GetUserInteractions(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
```
documentation; line too long.

jaedoucette replied:
Here, at least explain what is meant by mock function.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+
+// GetUserInteractions provides a mock function with given fields: uid, teamId, learningGroupPrefixes
+func (_m *UserStore) GetUserInteractions(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
+	ret := _m.Called()
```
The names here could be more self-documenting. I'm not sure what `ret` is.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+func (_m *UserStore) GetUserInteractions(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
+	ret := _m.Called()
+
+	var r0 store.StoreChannel
```
`r0` is only used as a placeholder for a return value. Why not omit it and have:

`if ...`
`    return rf()`
`if ret.Get(0) ...`
`    return ret.Get()...`
`return nil`

That would save some brackets and a local variable.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+}
+
+// GetUserTeams provides a mock function with given fields: uid, teamId, learningGroupPrefixes
+func (_m *UserStore) GetUserLearningGroups(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
```
Need more documentation on this function declaration.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+
+// GetUserTeams provides a mock function with given fields: uid, teamId, learningGroupPrefixes
+func (_m *UserStore) GetUserLearningGroups(uid string, teamId string, learningGroupPrefixes []*model.LearningGroupPrefix) store.StoreChannel {
+	ret := _m.Called()
```
This looks like it is the same function as above with a different name. Consider either making a function factory for these, or by making an alias for the function?

jaedoucette reviewed (commented):
I left a lot of comments. I think the main issues are:

- Documentation frequently just repeats what the code says, but in English. I'd like to see documentation that tells me what the code meant to be used for, why it was created, and what the intention of logical groups of lines is, rather than what each line does individually. Functions should ideally have auto-doc compatible comments so that IDEs can pull the documentation.

- The big in-line SQL queries seem pretty brittle and hard to read to me. I think you should use a Query builder to create the queries. A query builder is a library of functions that generates a query, but that produce a very readable, logical, syntax when the calls are written out. This would also help to separate the business logic from the SQL query, at least somewhat.

- It looks like there are a lot of little style issues that a linter would catch. I think you should adopt a GoLang linter and then address its comments. Using a Golang styler beforehand could also help.

A lot of my comments are repetitive things, so sorry about the volume!

mlippert commented:
_\- details in PR desc_ is very unhelpful in the PR summary. There should always be details in the description. I'm OK w/ the summary being _Add support for UserAnalytics_ and _Implement UserAnalytics_ by itself would also be fine.

mlippert commented:
@jaedoucette  go enforces its own concept of style (one I'm not fond of but it does provide consistency). And the base mattermost repos also always have a `check-style` target in the Makefile.

Running `make check-style` on the `develop` branch passes.

I just tried running `make check-style` on this branch and it failed. I can't say that the error is immediately clear, but I leave it to @brecriffs to figure out this (and any subsequent errors) and fix them.

brecriffs commented on api4/api.go:
```diff
@@ -131,6 +132,7 @@ func Init(a *app.App, root *mux.Router) *API {
 	api.BaseRoutes.User = api.BaseRoutes.ApiRoot.PathPrefix("/users/{user_id:[A-Za-z0-9]+}").Subrouter()
 	api.BaseRoutes.UserByUsername = api.BaseRoutes.Users.PathPrefix("/username/{username:[A-Za-z0-9\\_\\-\\.]+}").Subrouter()
 	api.BaseRoutes.UserByEmail = api.BaseRoutes.Users.PathPrefix("/email/{email}").Subrouter()
+	api.BaseRoutes.UserAnalytics = api.BaseRoutes.Users.PathPrefix("/{user_id:[A-Za-z0-9]+}/useranalytics").Subrouter()
```
Use user route here and add teamid as query param. ex
`api.BaseRoutes.User.PathPrefix("/teams/{team_id:[A-Za-z0-9]+}/useranalytics").Subrouter()`

brecriffs replied:
Completed.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,45 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+func (e *EdxLMS) GetLearningGroupPrefixes() ([]*LearningGroupPrefix, *AppError) {
+	var personalChannelNames []*LearningGroupPrefix = make([]*LearningGroupPrefix, len(e.PersonalChannels.ChannelList))
+
+	// Loop over PersonalChannels.ChannelList in the active LMS configuration and, for every key,
```
This is better, but I'd still rather see a description of _why_ we're doing this than a description of _what_ we're doing here. The what is fairly easy to understand from the code, but the why has to be guessed at.

brecriffs replied:
Ok I added another line in the comment, let me know if that is the 'why' you are looking for

jaedoucette replied:
Also, now that I've learned a little more Go, could you not write:

`for personalChannelName := range e.PersonalChannels.ChannelList {`
`     append(personalChannelNames, &LearningGroupPrefix{Prefix: personalChannelName})`
`}`

That's cleaner and saves the local variable. If you like, you can also then omit the `make(...)`  call on line 88 because go's append will automatically handle amortized growth.

jaedoucette replied:
> Ok I added another line in the comment, let me know if that is the 'why' you are looking for

Closer. I've dug in a little more now and I think I understand what this is for. How about:

"Construct an slice containing pointers to the prefixes of the names of each channel that this user is a part of."

I _think_ that's what we're really trying to do here.

brecriffs replied:
Your not really getting anything for this specific user, but I like where you went with it. I altered it to:

> Construct a slice containing pointers to the prefixes of the names of private channels that are for learning groups in this LMS

jaedoucette replied:
That's exactly the kind of info I'm looking for!

mlippert replied:
I don't think a comment describing the type of data such as "slice containing pointers" is a particularly good comment because it's a comment stating exactly what the code does.

Comments in the code can explain unusual constructs, or broad steps of the algorithm being used to accomplish what the function is supposed to accomplish
```
// Find the LMS for the active team
...
// Get the list of personal channel names from that LMS because those
// are the prefixes used for learning groups
...
// Create a learning group type for each channel name setting the prefix to that name
...
```
things like that. Where each of the `...` 's is the block of code to accomplish that comment.

jaedoucette replied:
@mlippert I like the idea of dropping the "slice containing pointers". Your documentation is significantly nicer for this.

brecriffs commented:
@mlippert I think this is ready for you to take a second look. I pushed a commit to mattermost-redux labelled 'temp commit'. It is, only temporary, but might be useful to check that changes are reflected correctly in both repos. I will be polishing that PR later today.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,45 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+
+	// Assert personalChannelNames as type 'array of pointers to LearningGroupPrefix structs'
+	var assertionSuccess bool
+	if personalChannelNames, assertionSuccess = interface{}(personalChannelNames).([]*LearningGroupPrefix); !assertionSuccess {
```
Is this cast actually needed? Since `personalChannelNames` was declared to be of this type, shouldn't the assertion always succeed?

brecriffs replied:
Well you are manipulating the struct in a way that may or may not change the dynamic type of the variable. I am a little rusty on this terminology (dynamic and static types) but I wanted to be sure here.

brecriffs replied:
From stackoverflow:

> A dynamic type means that it can hold a reference to different types (e.g. string, int, ...) and that it can change at runtime, whereas a static type is checked at compile time and cannot change.

jaedoucette replied:
NVM. Forgot that this is idiomatic check for nil in Golang.

jaedoucette commented on model/edx.go:
```diff
@@ -67,6 +67,45 @@ func (e *EdxLMS) GetOAuthConsumerSecret() string {
+}
+
+func (e *EdxLMS) CheckIfActiveLMS(currentTeamName string) bool {
+	// For every property of Teams in this learning management service (LMS) object (passed as e)
```
Sorry to keep adding review comments! 

Looking at this function again, I think I didn't think about what it was doing the first time round (another illustration of the importance of "why" comments!).

I _think_ what this function is doing is detecting whether `currentTeamName` matches any of the names of the teams stored in this EdxLMS structure. I _think_ the idea is that, if it does, then the passed team name must correspond to an "active" team. 

If that's so, I think it might be preferable to remove the `GetTeams()` function. It has a naming conflict with `GetTeams([]*model.Team, *model.AppError)` in `app/plugin_api.go`, and you're only using it here on line 78, when you could equally well write e.Teams (since you're still in a EdxLMS method). 

The comment also seems misleading or confusing to me. It is a bit tricky for me to tell what the `e.Teams` map is actually storing, but I think it maps context Ids to team names? If this is so, it might be clearer to say that we're going to iterate over all the names of the teams associated with this LMS. If not, it might be clearer to document what exactly that mapping is (properties is kinda vague).

brecriffs replied:
Just added more/revised documentation this function, let me know what you think. Also removed GetTeams, nice catch. I was using GetTeams from outside this package before, but now it makes sense just to use e.Teams.

brecriffs replied:
No worries on the comments, let's get this thing right! (Also, I'm learning how to write better code)

jaedoucette replied:
This is definitely clearer, but there's kinda of a lot of "what" vs. "why" in the comments. Maybe:

//This function determines whether this LMS object (e) is active.
//An LMS is active if one of the team names it contains matches the team name of the current user.
//The team name of the current user is passed as a string, and compared to the names of all the teams stored in this LMS.

brecriffs replied:
I like it, but changed it slightly to:

// This function determines whether this LMS object (e) is active
// An LMS is active if one of the team names it contains matches the team name of the currently selected MM team
// The team name of the currently selected MM team is passed as a string (currentTeamName), and compared to the names of all the teams stored in this LMS

What do you think?

mlippert replied:
I renamed this function to `IsLMSForTeam` and updated the function comment and implementation somewhat and pushed it for Brec a little while ago (I DM'd Bred so he knows about it) to look over, which may make this comment thread obsolete. (I also updated some of the rest of the LMS method changes)

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+	var r0 store.StoreChannel
+	if rf, ok := ret.Get(0).(func() store.StoreChannel); ok {
+		r0 = rf()
+	} else {
```
I think here you can use `else if {` instead of `else { if {`.

brecriffs replied:
I agree, but this syntax is all over the file. I vote to keep it as is.

jaedoucette replied:
Good point, I keep missing the local conventions.

jaedoucette commented on store/storetest/mocks/UserStore.go:
```diff
@@ -848,3 +848,35 @@ func (_m *UserStore) VerifyEmail(userId string) store.StoreChannel {
+	if rf, ok := ret.Get(0).(func() store.StoreChannel); ok {
+		r0 = rf()
+	} else {
+		if ret.Get(0) != nil {
```
same `else if {` thing here.

brecriffs replied:
"

jaedoucette commented on app/user.go:
```diff
@@ -1634,3 +1634,118 @@ func (a *App) RecordUserTermsOfServiceAction(userId, termsOfServiceId string, ac
+}
+
+// This function is used to fetch the learning group prefixes in the configuration file
+// for the selected mattermost team
```
Maybe add just a little more here. Why might you want to do this? I _think_ the reason is that you'll use these prefixes to construct an SQL query to build the network graph. The reason I think this is that I grep'd through the source code a couple of times to find where this was called, and then to find the functions that its results were passed to. If you document this fact here, a future reader will know what they might break (and what to re-test) if they change this.

jaedoucette replied:
NVM, I see that this is well documented below.

brecriffs replied:
Improved documentation.

brecriffs replied:
Oops, well improved it anyway, let me know if you think it's redundant.

jaedoucette replied:
I think the new stuff helps too. Now it's all very clear.

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+	// Denotes whether or not the current user was the recipient of this interaction
+	Is_Recipient *bool `json:"Is_Recipient"`
+
+	// The context in which this interaction occured (either 'Course' or the learning group type's prefix (ex. 'plg'))
```
typo _occured_ s/b _occurred_

brecriffs replied:
Fixed.

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+
+	// The name of the channel that this interaction took place in
+	Channel_Name *string `json:"Channel_Name"`
+}
```
- Why are the fields named with underscores rather than just PascalCase as seems to be the convention for everything else?
- Is the Channel_Name the display name or the slug name of the channel
- The comment says that the fields are pointers to allow for null values, but why do we want to allow null values for **all** of these fields? Looking at them I would argue that none of these fields should ever be null.

BTW I don't know how you discovered adding the tags to the struct field names for json, but that's excellent even if redundant as currently defined.

brecriffs replied:
Just pushed this commit: 

> Address some issues with models in model/user.go

    1. Changed field names to use PascalCase as opposed to with underscores (and changed sql queries to reflect changes)
    2. Updated comment regarding ChannelName (it's the channel's slug name)
    3. Only used pointers for values that could reasonably be null
    4. added json field names for all properties in each model, and followed conventions of other functions

mlippert replied:
I just looked at the changes in that commit, looks terrific!

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+	UserLearningGroups []*LearningGroup
+	LearningGroupTypes []*LearningGroupType
+	// Add more datasets here
+}
```
I think I would recommend that you add a field to contain what you refer to as the _current user_ id. I might name it _TargetUser_ I recommend this because the UserInteraction and LearningGroup types contain fields that are predicated on knowing what the current user is. By adding it here, we could get UserAnalytics for 2 different _target users_ and then be able to display info about both of them. Mostly though it lets the struct be self-contained and not require outside information in order to understand.

brecriffs replied:
Added TargetUser as a property of UserAnalytics

brecriffs replied:
Note that UserAnalytics is the struct that gets returned to redux and it may be redundant, if only for now, to store the TargetUser in redux, since we should know the userId that we sent. However, it might make sense to store it along with the rest of the user analytics data anyway, in case we are fetching for more than one user in the future.

mlippert replied:
Yes, I realized when I wrote this that the caller "knows" what the target user is, but I think adding it to the structure makes it more self-contained which I can see being useful. Thanks for adding it.

brecriffs replied:
Yeah makes sense. We can deal with it on the redux end now! I didn't include it in the list of datasets, but it will get returned alongside with the other datasets. It's possible that we would want: 

```
users: {
    analytics: [
         {
          TargetUser: userId1,
          UserInteractions: {},
          etc: {}
          },
          {
          TargetUser: userId2,
          UserInteractions: {},
          etc: {}
          },
     ]
}
```

Right now it is (in the current mattermost-redux PR):
```
users: {
    analytics: {
         TargetUser: userId1,
          UserInteractions: {},
          etc: {}
     }
}
```

brecriffs replied:
But we can discuss that on the other PR

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+	Channel_Name *string
+
+	// A comma delimited list of the usernames of the members of this learning group (channel)
+	Members *string
```
Why is this a string instead of an array(slice)?

brecriffs replied:
The sql query produces a comma delimited list, and this struct is for every row in the sql query. I believe I parse it into an array on the front-end, but it may be better to do that on the backend, although it would require some work.

mlippert commented on store/store.go:
```diff
@@ -283,6 +283,8 @@ type UserStore interface {
 	InferSystemInstallDate() StoreChannel
 	GetAllAfter(limit int, afterId string) StoreChannel
+	GetUserInteractions(uid string, teamId string, learningGroupTypes []*model.LearningGroupType) StoreChannel
+	GetUserLearningGroups(uid string, teamId string, learningGroupTypes []*model.LearningGroupType) StoreChannel
```
`uid` should be changed to `userId` for these 2 function definitions

brecriffs replied:
Changed.

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+	//     O = public channel
+	//     D = direct message (1 on 1)
+	//     G = group direct message
+	Channel_Type *string `json:"Channel_Type"`
```
The comment is is super helpful! Anything I might have thought about this Channel_Type fields contents was not what it is, but having read the comment, I can easily carry that knowledge around when looking at other code using it.

mlippert commented on app/user_test.go:
```diff
@@ -530,10 +530,10 @@ func TestRecordUserTermsOfServiceAction(t *testing.T) {
+		Email:       strings.ToLower(model.NewId()) + "success+test@example.com",
+		Nickname:    "Luke Skywalker", // trying to bring balance to the "Force", one test user at a time
+		Username:    "luke" + model.NewId(),
+		Password:    "passwd1",
```
Are these spacing changes because of errors reported by `check-style`?
If not I'd suggest reverting them because there are no other changes needed to this file for this branch.

brecriffs replied:
It was either check-style (which runs go fmt), or an explicit call to go fmt. In any case, go fmt automatically formats some things sometimes. Would you prefer to not commit these changes?

mlippert replied:
hmm, since the change were to clean up style errors that were reported, I think we can leave the changes. I'm guessing it was this file and 1 other.

mlippert commented on config/config-dev.json:
```diff
@@ -235,6 +235,106 @@
+                "Type": "edx"
+            }
+        ]
+    },
```
I think this LTISettings section should not be committed. It is good for your testing, and you probably should keep it in your local config.json, but the setting are very fake and somewhat invalid and not something that I think should be distributed.

brecriffs replied:
Sorry, didn't mean to commit that.

mlippert commented on store/sqlstore/user_store.go:
```diff
@@ -1361,3 +1361,470 @@ func (us SqlUserStore) InferSystemInstallDate() store.StoreChannel {
+
+// This function will run an sql query and return all relevant user interactions for the current user in the current mattermost team
+// Each result row will be of type UserInteraction (defined in model/user.go)
+// TO-DO: Break the query in this function into smaller, more maintainable queries perhaps using a query builder
```
Good comment but change `TO-DO` to `TODO`. it's just how it is typically written. grep the code base for TODO you'll see a bunch of them. (Mostly in all of the go packages, but several in the mattermost code)

brecriffs replied:
Changed.

mlippert commented on api4/user.go:
```diff
@@ -15,6 +15,7 @@ import (
 	"github.com/mattermost/mattermost-server/mlog"
 	"github.com/mattermost/mattermost-server/model"
 	"github.com/mattermost/mattermost-server/store"
+	"github.com/mattermost/mattermost-server/web"
```
OK, I changed my mind about the changes you made that were for things reported by `go fmt` but not by `make check-style`. In particular because of the extensive changes to this file. When I thought it was just some small spacing changes in a couple of files, I didn't think much of it, but there are a lot of lines changed in this file, from a commit (0312a3a581845abdce9e9b03af46f55679f0bd4f) which said it was because of fixing things reported by `check-style` but probably wasn't since `make check-style` passed w/o any errors on the develop branch.

Anyway the result is lets not make unnecessary changes to this code base.

brecriffs replied:
Reverted.

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+}
+
+// This is a struct for all user analytics data sets
+// This is the struct that gets returned to mattermost-redux in the users/{userId}/teams/{teamId}/useranalytics request
```
Another note, mattermost-redux is a client of the mattermost-server api but it is not the only client. The api calls can be used w/ anyone w/ the proper url (and hopefully authorization), so mentions mattermost-redux as the specific recipient is probably wrong.
That said, I don't care enough to insist it be changed, but it was something I felt should be said so you think about it in the future.

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+
+	// The type of learning group (personal channel name from the config)
+	//     ex. plg
+	LearningGroupType string `json:"learning_group_type"`
```
This maybe should be a LearningGroupType, but right now it is just the LearningGroupType's prefix. I sort of feel that either the name or the type should be changed, because otherwise it's confusing. I think just changing the name is reasonable, back to LearningGroupPrefix which can be used to find the LearningGroupType in the array of them that is returned if that was desired (say to get other properties, which might exist in the future)

mlippert commented on model/user.go:
```diff
@@ -47,6 +47,77 @@ const (
+	ChannelName string `json:"channel_name"`
+
+	// A comma delimited list of the usernames of the members of this learning group (channel)
+	Members *string `json:"members"`
```
I feel like there was a comment/discussion on this before, that Member should be a slice? But I don't see it.
It is [really easy](https://yourbasic.org/golang/split-string-into-slice/) to do in go and I think it makes sense to do it here rather than returning a string and having the recipient process it into an array later.
```
s := strings.Split("a,b,c", ",")
fmt.Println(s)
// Output: [a b c]
```
and define
```
Members []string `json:"members"`
```
I see that to actually accomplish this you'd need to transform the result of the query in the `store/sqlstore/user_store.go` to this format. I guess we can hold off on that for now because it's a little complicated, but I do think it would be better long term.

brecriffs committed:
Implemented new API to get UserAnalytics

The capability to get different types of user analytics exists by
supplying different names to the call. The call requires the team id,
the user id and the analytic name.

User analytics is focused on support of the use of RiffEdu mattermost
with an LMS and LTI sign-on. As such the current data returned is for
'UserInteractions' and in particular interactions w/ the student's "teams"
defined by the members of the private channels they are assigned to via
the LTISettings config for sign-on from that LMS.

mlippert reviewed (approved):
This looks good. :+1:

mlippert commented:
Squashed the commits in this branch into a single commit now that it is approved and force pushed it.

mlippert commented:
**Release** [1.0.0-dev.9](https://github.com/rifflearning/mattermost-server/releases/tag/1.0.0-dev.9)
